src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+	char	*f2 = NULL;
+	char	*fname = NULL;
+	char	*cont_name = NULL;
+	int	rc = 0;
+
+	if (path == NULL || _obj_name == NULL || _cont_name == NULL)
+		return -EINVAL;
+
+	if (strcmp(path, "/") == 0) {
+		*_cont_name = strdup("/");
+		if (*_cont_name == NULL)
+			return -ENOMEM;
+		*_obj_name = NULL;
+		return 0;
+	}
+
+	f1 = strdup(path);
+	if (f1 == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	f2 = strdup(path);
+	if (f2 == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	fname = basename(f1);
+	cont_name = dirname(f2);
+
+	if (cont_name[0] == '.' || cont_name[0] != '/') {
+		char cwd[1024];
+
+		if (getcwd(cwd, 1024) == NULL) {
+			rc = -ENOMEM;
+			goto out;
+		}
+
+		if (strcmp(cont_name, ".") == 0) {
+			cont_name = strdup(cwd);
+			if (cont_name == NULL) {
+				rc = -ENOMEM;
+				goto out;
+			}
+		} else {
+			char *new_dir = calloc(strlen(cwd) + strlen(cont_name)
+					       + 1, sizeof(char));
+			if (new_dir == NULL) {
+				rc = -ENOMEM;
+				goto out;
+			}
+
+			strcpy(new_dir, cwd);
+			if (cont_name[0] == '.') {
+				strcat(new_dir, &cont_name[1]);
+			} else {
+				strcat(new_dir, "/");
+				strcat(new_dir, cont_name);
+			}
+			cont_name = new_dir;
+		}
+		*_cont_name = cont_name;
+	} else {
+		*_cont_name = strdup(cont_name);
+		if (*_cont_name == NULL) {
+			rc = -ENOMEM;
+			goto out;
+		}
+	}
+
+	*_obj_name = strdup(fname);
+	if (*_obj_name == NULL) {
+		free(*_cont_name);
+		*_cont_name = NULL;
+		rc = -ENOMEM;
+		goto out;
+	}
+
+out:
+	if (f1)
+		free(f1);
+	if (f2)
+		free(f2);
+	return rc;
+}
+
+/**
+ * Parse path into sys_path->dir_name and sys_path->name.
+ */
+static int
+dfs_sys_path_parse(dfs_sys_path_t *sys_path, const char *path)
+{
+	/**
+	 * TODO integrate with parse_filename so there is only one function.
+	 */
+	return parse_filename(path, &sys_path->name, &sys_path->dir_name);
+}
+
+/**
+ * Free a dfs_sys_path_t.
+ */
+static void
+dfs_sys_path_free(dfs_sys_path_t *sys_path)
+{
+	free(sys_path->dir_name);
+	free(sys_path->name);
+
+	/**
+	 * TODO Decrement reference to parent.
+	 */
+}
+
+/**
+ * Initialize dfs_sys_path_t.
+ */
+static int
+dfs_sys_path_init(dfs_sys_t *dfs_sys, dfs_sys_path_t *sys_path,
+		  const char *path)
+{
+	int rc;
+
+	rc = dfs_sys_path_parse(sys_path, path);
+	if (rc)
+		return rc;
+
+	rc = dfs_sys_hash_lookup(path, dfs_sys, &sys_path->parent);
+	if (rc) {
+		dfs_sys_path_free(sys_path);
+		return rc;
+	}
+
+	/**
+	 * Handle the case of root "/".
+	 * TODO double check this
+	 */
+	if (sys_path->name == NULL) {
+		sys_path->parent = NULL; /* TODO set to root explicitly? */
+		sys_path->name = sys_path->dir_name;
+		sys_path->dir_name = NULL;
+	}
+
+	return rc;
+}
+
+/**
+ * The calls below here should be as similar as possible to the
+ * POSIX versions, but with a dfs_sys_t parameter.
+ *
+ * TODO Do we want these to set errno and return -1, like POSIX,
+ * or do we want them to return errno, like DFS?
+ */
+int
+dfs_sys_access(dfs_sys_t *dfs_sys, const char* path, int amode)
```

Comment:
What guarantees does this function give as the path afterwards?  Can it modify it, is it safe for the caller to use it afterwards, what if it's in read-only memory?

This function should also take a len for path as well.

Reply:
> What guarantees does this function give as the path afterwards? Can it modify it, is it safe for the caller to use it afterwards, what if it's in read-only memory?

The path is `const char*`, and it shall not be modified by this function.



> This function should also take a len for path as well.

I think we need to think about this one, because it would imply that ALL of the wrappers should take a `len`, but the DFS API doesn't take a `len` - consider `dfs_lookup`, for example.
Anyway, it's not difficult to do so.